### PR TITLE
Use model class DB connection in have_db_index matcher

### PR DESCRIPTION
**Pull request by jooeycheng**
Fix #1051. This PR is WIP. 

Hey @mcmire, lemme know what you think.

I'm thinking to rename `ProductionRecord` to something else, as this naming might cause confusion.

### CODE REVIEW

File: spec/support/unit/helpers/model_builder.rb [hunk 1]
```diff
@@ -35,7 +35,7 @@ def reset
       end
 
       def create_table(table_name, options = {}, &block)
-        connection = ::ActiveRecord::Base.connection
+        connection = options.delete(:connection) || ::ActiveRecord::Base.connection
```
[the remark this answers is not in the document]
Comment (jooeycheng) [reply spec/support/unit/helpers/model_builder.rb]: Fixed.

File: spec/support/tests/database_configuration.rb [hunk 2]
```diff
@@ -3,7 +3,7 @@
 
 module Tests
   class DatabaseConfiguration < SimpleDelegator
-    ENVIRONMENTS = %w(development test production)
+    ENVIRONMENTS = %w(development production)
```
[the remark this answers is not in the document]
Comment (jooeycheng) [reply spec/support/tests/database_configuration.rb]: Reverted changes on this line.

File: spec/unit/shoulda/matchers/active_record/have_db_index_matcher_spec.rb [hunk 3]
```diff
@@ -3,7 +3,8 @@
 describe Shoulda::Matchers::ActiveRecord::HaveDbIndexMatcher, type: :model do
   context 'have_db_index' do
     it 'accepts an existing index' do
-      expect(with_index_on(:age)).to have_db_index(:age)
+      expect(with_index_on(:age1)).to have_db_index(:age1)
+      expect(with_index_on(:age2, parent_class: ProductionRecord)).to have_db_index(:age2)
```
[the remark this answers is not in the document]
Comment (jooeycheng) [reply spec/unit/shoulda/matchers/active_record/have_db_index_matcher_spec.rb]: That makes sense. I have updated accordingly.

**Comment by jooeycheng** [comment]
There're still some failing specs. I'll look into those over the weekend.

**Comment by mcmire** [comment]
@jooeycheng I looked into this a bit for you. I think the problem is that both of the models you're creating in that test have the same class name. The method that creates classes will (unhelpfully, now, I realize) detect whether the class it's creating already exists and remove it before adding its version. After the tests are run, it will then make sure that class is removed. So the second model is overwriting the first model, however, when it goes to remove that class after the test is finished running, it gets very confused. If you give `with_index_on` the ability to specify a class name and then pass that down, I think it should work a lot better.

**Comment by jooeycheng** [comment]
```ruby
```

**Comment by jooeycheng** [comment]
@mcmire Ahh, that makes sense. I have updated `with_index_on` to accept table_name instead of model_name, bcos i found that converting from snake_case to PascalCase is easier than PascalCase to snake_case.

```ruby
"employees".classify
# => Employee

# Notice it's smart enough to remove the trailing "s"
```

```ruby
"Employee".underscore
# => "employee"
"Employees".underscore
# => "employees"

# Well, I just don't know an easy way to convert PascalCase to snake_case
```

**Comment by mcmire** [comment]
@jooeycheng Oops, I noticed you updated this just now. Skimming over it again I think it looks good, but I'll take a closer look when I get a chance.

**Comment by mcmire** [comment]
Thanks so much for this!

**Comment by jooeycheng** [comment]
@mcmire Thanks for taking the time to guide me through it 🙆‍♂️. Your [comment](https://github.com/thoughtbot/shoulda-matchers/pull/1051#issuecomment-468949085) was super helpful, and I've really learnt a lot from reading the codebase. I'm impressed that it creates a Rails application on-the-fly in test.

**Comment by mcmire** [comment]
@jooeycheng Yeah there's definitely some complex parts to the gem that not many people know about but I'm glad I was able to help you navigate them. If you find something else let me know and I will happily take another PR of yours :)

**Pull request merged**